offchain/danaswap-browser/src/DanaSwap/Browser/Home.purs

```diff
@@ -19,18 +24,239 @@ main = HA.runHalogenAff do
+                    ]
+                ]
+            , HH.div [ mkClass "column" ] [ HH.text " " ]
+            ]
+
+        ]
+      <> concat
+        ( renderPool
+            <$> pools
+        )
+    where
+    renderPool pool@(Pool p) =
+      [ HH.div
+          [ mkClass "pools-table-row columns is-mobile is-vcentered p-3 my-3"
+          , HE.onClick $ \_ ->
+              if currentPool == Just pool then
+                ResetCurrentPool
+              else
+                SetCurrentPool pool
+          ]
+
+          [ HH.div [ mkClass "column" ]
+              [ HH.p [ mkClass "is-size-5-desktop is-size-7-touch has-text-left" ] [ HH.text $ show pool ]
+              ]
+          , HH.div [ mkClass "column" ]
+              [ HH.p [ mkClass "is-size-5-desktop is-size-7-touch has-text-centered" ] [ HH.text $ show p.assetA <> " / " <> show p.assetB ]
+              ]
+          , HH.div [ mkClass "column" ]
+              [ HH.p [ mkClass "is-size-5-desktop is-size-7-touch has-text-centered" ] [ HH.text $ show p.fee <> "ADA" ]
+              ]
+          , HH.div [ mkClass "column has-text-centered" ]
+              [ if currentPool == Just pool then mkIcon "angle-up" else mkIcon "angle-down"
+              ]
+          ]
+      ] <> renderPoolDropdown pool
+
+    renderPoolDropdown pool =
+      if currentPool == Just pool then
+        [ HH.div [ mkClass "pools-table-row-dropdown box my-6" ]
+            [ HH.div [ mkClass "columns is-mobile" ]
+                [ HH.div [ mkClass "column is-narrow" ]
+                    [ HH.button [ HE.onClick $ const (SetActivity Swap), mkClass $ "button is-rounded is-medium is-responsive " <> if (activity == Swap) then "danaswap-btn-has-background" else "danaswap-btn-has-border" ] [ HH.text "SWAP" ]
+                    ]
+                , HH.div [ mkClass "column is-narrow" ]
+                    [ HH.button [ HE.onClick $ const (SetActivity AddLiquidity), mkClass $ "button is-rounded is-medium is-responsive " <> if (activity == AddLiquidity) then "danaswap-btn-has-background" else "danaswap-btn-has-border" ] [ HH.text "ADD LIQUIDITY" ]
+                    ]
+                , HH.div [ mkClass "column is-narrow" ]
+                    [ HH.button [ HE.onClick $ const (SetActivity WithdrawLiquidity), mkClass $ "button is-rounded is-medium is-responsive " <> if (activity == WithdrawLiquidity) then "danaswap-btn-has-background" else "danaswap-btn-has-border" ] [ HH.text "WITHDRAW" ]
+                    ]
+                ]
+            , case activity of
+                Swap -> renderSwapForm pool
+                AddLiquidity -> renderAddLiquidityForm pool
+                WithdrawLiquidity -> renderWithdrawLiquidityForm
+            ]
+        ]
+      else []
+
+    renderSwapForm (Pool ({ assetA: (Asset a), assetB: (Asset b) })) = HH.div [ mkClass "mt-6" ]
+      [ HH.div [ mkClass "columns is-mobile is-vcentered" ]
```

Comment:
You have `<input>` elements, but they aren't in a `<form>` ?? :thinking:

Reply:
these inputs will help to construct the state directly, that's why there's no form wrapping them